### ISSUE TITLE
Gauge is crashing with OOM when running 50 Specs in parallel

### ISSUE THREAD

**Issue by saritapol13** [getgauge/gauge#492]
[GaugeCrashLogs.txt](https://github.com/getgauge/gauge/files/390697/GaugeCrashLogs.txt)
### Expected behavior

Gauge specification should run in parallel without issues
### Actual behavior

Gauge crashes when running in 50 specification in parallel
### Steps to reproduce

Run 50 specs in parallel

Executing in 50 parallel streams.

[runner: 45] Error occurred during initialization of VM
java.lang.OutOfMemoryError[runner: 45] : unable to create new native thread
[runner: 4] Error occurred during initialization of VM
[runner: 4] java.lang.OutOfMemoryError: unable to create new native thread
runtime: failed to create new OS thread (have 100 already; errno=11)
fatal error: newosproc
[GaugeCrashLogs.txt](https://github.com/getgauge/gauge/files/390687/GaugeCrashLogs.txt)

Machine on which Tests are running(Linux OS):
Linux  3.8.13-55.2.1.el6uek.x86_64 #2 SMP Wed Jul 15 14:19:51 PDT 2015 x86_64 x86_64 x86_64 GNU/Linux

Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                2
On-line CPU(s) list:   0,1
Thread(s) per core:    1
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 26
Model name:            Intel(R) Xeon(R) CPU           E5530  @ 2.40GHz
Stepping:              4
CPU MHz:               2400.085
BogoMIPS:              4800.17
Hypervisor vendor:     VMware
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              8192K
NUMA node0 CPU(s):     0,1
### Gauge version

Gauge version: 0.6.0
## Plugins

html-report (2.1.1)
java (0.5.0)

**Comment by saritapol13** [comment]
Any updates on this?

**Comment by apoorvam** [comment]
@saritapol13 Apologies for the late response, the team was held up with some other tasks. 

From the logs, it looks like you are trying to execute 50 specs, each in a parallel stream i.e. 50 parallel streams. As documented [here](http://getgauge.io/documentation/user/current/advanced_readings/execution_types/parallel_execution.html), the number of parallel streams(--n flag) should be specified depending on CPU cores available, beyond which it may fail. Given that you are using 2 core machine, its not ideal to parallelize in 50 streams. This is because: In the case of Java, 50 streams will spawn 50 java runners, which in turn could be spinning number of Java processes/threads. The system may not be able to handle this load due to memory issues. 

Can you try parallelizing in 2 or 4 streams as `gauge -p --n=4 specs`? This will run all your specs in 4 parallel streams and generates report seamlessly. If this works, try doing it with more parallel streams. 

Let us know how it goes.

**Comment by saritapol13** [comment]
Hi Apporvam,

We tried running in different configurations. We wanted to maintain same number of parallel execution as we did for Twist.

On linux machine with 2 cores and 8 GB ram we were able to run almost 200 twist tests in parallel.
Now with gauge we can at max run 30 java processes with 4 cores and 8 GB ram. Its impacting badly on execution time of the regressions. They have almost doubled , tripled the regression run timings.

Can you please see if we can fix something as part of the memory consumed by each java runner? Because its consuming almost 5 MB per java process/runner which I think is concern as part of performance or resource usage point of view.

From the end users point of view this is not justifying when we want to run specs in parallel. Running just 20-30 specs is not a good deal as compared to Twist ant support where in we could run 200 tests with the same machine config

Please let me know if you need any additional information

Thanks,
Sarita

**Comment by sriv** [comment]
@saritapol13 - Thank you for the analysis. Unfortunately Gauge's current design brings up a new JVM for every parallel stream. This was done to keep it consistent across all languages that Gauge supports (Some languages are single threaded).

We are working on an idea that could possibly optimize this. Will keep this issue posted of the same.

Thanks.

**Comment by saritapol13** [comment]
Thanks a lot @sriv for considering this issue

**Comment by kashishm** [comment]
@saritapol13 We have done a spike to run specs in parallel using threads instead of multiple JVM's.

Can you please try installing both gauge-java and gauge from `multithreading` branch source and then try executing specs in parallel.

The specs can fail, if the user code is not thread-safe. In that case, you might have to refactor your test implementation code.

**Comment by saritapol13** [comment]
@kashishm  sure I will try and update you about the status

**Comment by saritapol13** [comment]
Hi @kashishm 
I tested on windows and I still see high CPU usage. The one change I see is now single jvms are getting spawned on windows.

There are other issues I am facing, I see that the same spec is running on the number of parallel count that we specify. So if I specify n=3 same spec is running 3 times. I had added some loggers and see that same log is displayed 3 times for all the specs. So the run timings are increased to like 3 times. Not sure whats happening here. Until this is fixed I wont be able to find whether this has fixed the CPU usage and RAM usage because I wont be having accurate statistics.

Thanks,
Sarita

**Comment by saritapol13** [comment]
Hi @kashishm 

The scenario has passed still I get below error

Capturing regular screenshot..
[runner: 2]     Error Message: java.lang.NullPointerException
    Stacktrace: 
    gaugecore.utils.ScenarioContext.tearDown(ScenarioContext.java:179)
    sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke(Method.java:601)
    com.thoughtworks.gauge.execution.MethodExecutor.execute(MethodExecutor.java:38)
    com.thoughtworks.gauge.execution.HooksExecutor$TaggedHookExecutor.executeHook(HooksExecutor.java:101)
    com.thoughtworks.gauge.execution.HooksExecutor$TaggedHookExecutor.execute(HooksExecutor.java:87)
    com.thoughtworks.gauge.execution.HooksExecutor.execute(HooksExecutor.java:44)
    com.thoughtworks.gauge.processor.MethodExecutionMessageProcessor.executeHooks(MethodExecutionMessageProcessor.java:60)
    com.thoughtworks.gauge.processor.ScenarioExecutionEndingProcessor.process(ScenarioExecutionEndingProcessor.java:32)
    com.thoughtworks.gauge.connection.MessageDispatcher.dispatchMessages(MessageDispatcher.java:75)
    com.thoughtworks.gauge.GaugeRuntime$1.run(GaugeRuntime.java:70)
    java.lang.Thread.run(Thread.java:722)

Thanks,
Sarita

**Comment by saritapol13** [comment]
Hi, 

After lot of debugging figured out that the spec enter into BeforeScenario hook twice. I have tried running just single spec and n=1. 

Thanks,
Sarita

**Comment by kashishm** [comment]
@saritapol13 Thanks for the analysis. The spec is getting executed once but the execution hooks are executed n(number of streams)+1 times. This is happening because the [Hooks Registry](https://github.com/getgauge/gauge-java/blob/master/src/main/java/com/thoughtworks/gauge/registry/HooksRegistry.java#L35) in gauge-java is not thread-safe. We will update the issue after fixing this.

**Comment by kashishm** [comment]
@saritapol13 Can you try again with the latest gauge and gauge-java source?

**Comment by saritapol13** [comment]
@kashishm yes I will try again and update you the status.

**Comment by saritapol13** [comment]
@kashishm  I still see the same issue.

**Comment by saritapol13** [comment]
Hi @kashishm,

I performed the runs with Twist and Gauge on the same machines to have better comparison.

Good part is we dont see any out of memory happenings. 
CPU usage is high continuously and also memory usage.

The run timings are almost doubled in comparison with Twist runs.  We performed multiple runs and this behaviour is seen.

The best part if atleast we could manage to run the gauge runs on the same machine config which was not the case earlier.

Thanks,
Sarita

**Comment by saritapol13** [comment]
One more issue I saw was, DataStoreFactory.getScenarioDataStore() here the data is shared between different specs and the tests are failing because of this

**Comment by saritapol13** [comment]
Any updates on this?

**Comment by saritapol13** [comment]
Hi,

Please let me know when can we expect fix for this issue.

Thanks,
Sarita

**Comment by kashishm** [comment]
We did some performance analysis between Twist and Gauge execution time, we tried to execute different projects with **50 specs** with different numbers of execution streams. Here are the results.

|Tool|50Runners-16sec/spec|50Runners-40sec/spec|50Runners-240sec/spec|10Runners-240sec/spec|
|----|--------------------|---------------------|---------------------|---------------------|
|Twist|24sec|49sec|4min 09sec|20min 09sec|
|Before improvements in Gauge Java|49sec|1min 11sec|4min 34sec|20min 15sec|
|After improvements in Gauge Java|19sec|43sec|4min 3sec|20min 3sec|

The CPU utilization was high because it was scanning for step implementations and hooks in the classpath in every thread. Now it only scans it once. So, the CPU utilization is also improved.

@saritapol13 Please install Gauge and Gauge java from updated multithreading branch. Please share the execution stats and CPU usage for your project.

**Comment by saritapol13** [comment]
@kashishm Thank you for the performance stats shared. We will perform the runs with the latest code from the multithreading branch. We will keep the issue updated with the status.

**Comment by saritapol13** [comment]
The spec is getting executed once but the execution hooks are executed n(number of streams)+1 times again. I have taken the latest code for gauge and gauge-java

**Comment by kashishm** [comment]
@saritapol13 Is this happening with all the hooks or just suite level hooks? I see this happening only with `BeforeSuite` and `AfterSuite` hooks, other hooks are working as expected.

**Comment by saritapol13** [comment]
@kashishm There was issue in our test framework which is fixed. We just use @setup and @teardown hooks. So no issues with this hooks. Apologize for the wrong issue being reported.

When we give specs directory for running the specs whether specs from nested directory are also picked?

**Comment by kashishm** [comment]
@saritapol13 The specifications from the nested directories will also be included in the execution.

**Comment by saritapol13** [comment]
@kashishm the specs are not getting picked from the nested directories. Only direct specs from that directory are getting executed on linux machine. While the same code is working fine on Windows machine. I am also trying to debug more whats happening.  Gauge and Gauge-java are built from the multithreading branch code

**Comment by saritapol13** [comment]
@kashishm  I tried simple to execute from the command prompt on linux machine still the same issue (Only direct specs are executing and not from the nested directories)
gauge -n=4 --strategy="lazy" /tmp/users/specs

**Comment by kashishm** [comment]
@saritapol13 I cannot reproduce it. It is working as expected on mac and linux.

**Comment by kashishm** [comment]
@saritapol13 Can you share the performance/CPU/execution stats for your project after the performance improvements?

**Comment by saritapol13** [comment]
@kashishm yes I have started the runs. Once the results are available I will share the performance/cpu/execution stats.

**Comment by saritapol13** [comment]
@kashishm We are facing issue with html report plugin. Spec htmls are not getting created. https://github.com/getgauge/html-report/issues/120. I have total runtiming  but detailed spec information is missing and also its individual run time. I would need this to have more clarity on the defect

**Comment by sriv** [comment]
@saritapol13 - getgauge/html-report#120 is fixed and closed. Are you able to see the reports generated? If so could you please share some statistics comparing the run?

We are looking to bring this feature into master, once we know that this works well in your usecase. Thanks.

**Comment by saritapol13** [comment]
@sriv we are on it and performing the runs. 
The sad part is the runs which we are performing there are issues in our product and we are not having stable runs to have accurate comaprison.

Still I will share whatever runs we have performed so far

Job1:
![job1](https://cloud.githubusercontent.com/assets/12542343/21842308/29e2e618-d80c-11e6-9278-53c5b0790d1a.png)
![job2](https://cloud.githubusercontent.com/assets/12542343/21842345/5b9793f2-d80c-11e6-9a27-aaa604071c36.png)

**Comment by saritapol13** [comment]
I missed to add 1st line is of Twist and 2nd one is for Gauge. We see the improvements in runtime and also no spikes during that time. But as you see there are failed tests which might increase the run time once passed

**Comment by saritapol13** [comment]
Hi @apoorvam @sriv ,

We performed multiple runs and defect is fixed :) The run results of gauge in some cases are 20% better than Twist and in some case they match near by. They never exceed the time as compared to Twist which is very very best thing. 

Thank you for all the efforts and support, putting in all pieces together to make this achieve. 

If you need more stats we can add accordingly.

Thanks,
Sarita

![product1](https://cloud.githubusercontent.com/assets/12542343/22505889/99e98aa0-e8a3-11e6-9c51-142ec91046a5.PNG)

**Comment by saritapol13** [comment]
Hi,

We are seeing that the cache memory being used during the gauge run is not getting freed up after the run. This does impact when you want to perform next run on the same machine.

Thanks,
Sarita

**Comment by kashishm** [comment]
@saritapol13 We have merged the multithreading branch of `gauge` and `gauge-java` into master.
We have fixed some issues which might have impacted the execution after the first run. 
Could you please try `gauge` and `gauge-java` from `master` branch and let us know if you still face the same issue.

**Comment by kashishm** [comment]
To use multithreading in parallel execution, users will have to set `enable_multithreading` property/env var to true. Currently, only `gauge-java` language runner supports multithreading parallel execution.

**Comment by saritapol13** [comment]
@kashishm Thank you. We will test this on master version.

**Comment by saritapol13** [comment]
@kashishm Whether enable_multithreading variable value needs to be set in env\default\default.properties file?

**Comment by sguptatw** [comment]
@saritapol13 : Adding the property in env\default\default.properties will work if you are using the default env.
This property has to be set in the env used for execution.

**Comment by saritapol13** [comment]
@kashishm The value from the environment variable is not getting used and the original problem persists.

Is the variable name as multithreaded as referenced in java.json?

Gauge version: 0.8.0

Plugins
-------
html-report (3.1.0)
java (0.6.0)

**Comment by kashishm** [comment]
@saritapol13 Did you install `gauge` and `gauge-java` from `master` branch?

If you have installed from `master` branch then you should have `gauge` and `gauge-java` with the following versions
```
Gauge version: 0.8.1

Plugins
-------
java (0.6.1)
```
But I see in the above comment that you have released version of `gauge` and `gauge-java`
```
Gauge version: 0.8.0

Plugins
-------
java (0.6.0)
```

**Comment by saritapol13** [comment]
@kashishm Do you want me to build gauge and gauge-java from the master source code?

**Comment by saritapol13** [comment]
@kashishm  I built gauge from the source code but still I see version as 0.8.0

**Comment by kashishm** [comment]
@saritapol13 Any updates on this? Is the execution still slow after the first run?

**Comment by saritapol13** [comment]
@kashishm  We have performed only 1 run after the first run and we dont see issues about the cache. Still we are performing many runs to see whether the issue is resolved completely. We will keep the defect posted with the updates.

**Comment by sguptatw** [comment]
When the value is not set in the properties file to enable multithreading
Getting `Incorrect value for enable_multithreading in property file. Cannot convert to boolean.`
Can we have something like `multithreading is disabled as no property is set in the properties file`
instead of the previous message?

**Comment by sguptatw** [comment]
Is there a way to know that multithreading has been enabled by looking at the logs or the debug info or the reports.

**Comment by sguptatw** [comment]
mvn test -Dflags="--verbose" -DinParallel=true is not working as expected
The verbose mode of sequential execution gives a lot more information.

**Comment by sguptatw** [comment]
Getting `Failed to read proto message: unexpected EOF` when executing
`mvn test -DinParallel=true -Dnodes=100 -DspecsDir=specs/one`

The test suite has 81 specs and 25 scenarios each, however the specs/one folder has 1 spec with 25 scenarios in it.

**Comment by sriv** [comment]
@sguptatw - some answers inline:

> When the value is not set in the properties file to enable multithreading
Getting Incorrect value for enable_multithreading in property file. Cannot convert to boolean.
Can we have something like multithreading is disabled as no property is set in the properties file
instead of the previous message?

Yes, ideally if any property isn't set, `gauge` should fallback to using the default value. In this case `enable_multithreading=false` by default.

>Is there a way to know that multithreading has been enabled by looking at the console or the debug info or the reports?

No. Unless the test code writes it explicitly. Gauge does not print much contextual information ex parallel run mode, strategy used etc. This requirement could be extrapolated to a larger set. Please open a separate issue, if you wish for such information to be available in reports.

> mvn test -Dflags="--verbose" -DinParallel=true is not working as expected
The verbose mode of sequential execution gives a lot more information.

Please elaborate. This comment does not tell us what you expect and what is the actual outcome. Perhaps worth opening a separate issue.

> Getting Failed to read proto message: unexpected EOF when executing
mvn test -DinParallel=true -Dnodes=100 -DspecsDir=specs/one
The test suite has 81 specs and 25 scenarios each, however the specs/one folder has 1 spec with 25 scenarios in it.

Again, without steps to replicate it is difficult to comment. Please create an issue, and add more information.

**Comment by kashishm** [comment]
@saritapol13 We have released gauge([0.8.1](https://github.com/getgauge/gauge/releases/tag/v0.8.1)) and gauge-java([0.6.1](https://github.com/getgauge/gauge-java/releases/tag/v0.6.1)) with the ability to use multithreading in parallel execution. Please try it out and let us know if you face any issue.

**Comment by saritapol13** [comment]
@kashishm thank you for releasing the gauge version. We will update the status of the defects after the runs. We dont see issues accross the cache memory now. We will continue performing the runs.
